### PR TITLE
chore: swipe migrate v0.5.3 to v0.6.0

### PR DESCRIPTION
**Pull request by acud**
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/1716)

### CODE REVIEW

**Comment by Eknir** [comment]
I'm getting a panic here. 

```
INFO[2021-05-17T13:35:37+02:00] version: 0.6.0-637caccb                      
Password: 
WARN[2021-05-17T13:35:38+02:00] clef is not enabled; portability and security of your keys is sub optimal 
INFO[2021-05-17T13:35:38+02:00] using existing swarm network address: 25bc2c9f3dd3be55e9bf8b211a82f7434c9788cb847162a965b16e09a81fe156 
INFO[2021-05-17T13:35:38+02:00] swarm public key 03741675e5d524475d5f62908f3c7cab4b33e5fe92a0754e9f78f45176dd60f31c 
DEBU[2021-05-17T13:35:38+02:00] using existing libp2p key                    
DEBU[2021-05-17T13:35:38+02:00] using existing pss key                       
INFO[2021-05-17T13:35:38+02:00] pss public key 027cc5c79fcb5d63efb91f951e09cf48b9a61bc8b26d653f01e219163243f46ac8 
INFO[2021-05-17T13:35:38+02:00] using ethereum address a017448fdb75191e399c16fa7fd29bc9dc181e7e 
INFO[2021-05-17T13:35:38+02:00] using default factory address for chain id 5: f0277caffea72734853b834afc9892461ea18474 
INFO[2021-05-17T13:35:38+02:00] using existing chequebook d8a66097276f73bd9df51207342d371eae45ad92 
INFO[2021-05-17T13:35:39+02:00] database capacity: 5000000 chunks (approximately 19.5GB) 
INFO[2021-05-17T13:35:39+02:00] localstore migration: found current localstore schema code, migrate to yuj, total migrations 2 
INFO[2021-05-17T13:35:39+02:00] localstore migration: need to run 1 data migrations on localstore to schema code 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x98 pc=0xcaede2]

goroutine 1 [running]:
github.com/ethersphere/bee/pkg/shed.(*DB).NewIterator(0x0, 0x0, 0x0)
	github.com/ethersphere/bee/pkg/shed/db.go:161 +0x22
github.com/ethersphere/bee/pkg/shed.Index.Count(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/ethersphere/bee/pkg/shed/index.go:511 +0x53
github.com/ethersphere/bee/pkg/localstore.migrateYuj(0xc0021b4000, 0x15a4805, 0x4f)
	github.com/ethersphere/bee/pkg/localstore/migration.go:113 +0xb0
github.com/ethersphere/bee/pkg/localstore.(*DB).migrate(0xc0021b4000, 0xc00069e940, 0x4, 0x10, 0xc00069e940)
	github.com/ethersphere/bee/pkg/localstore/migration.go:58 +0x244
github.com/ethersphere/bee/pkg/localstore.New(0xc002192520, 0x1b, 0xc0001d0200, 0x20, 0x20, 0xc0006a9398, 0x1887e60, 0xc0002dc6c0, 0x154d468, 0x5, ...)
	github.com/ethersphere/bee/pkg/localstore/localstore.go:246 +0x17dc
github.com/ethersphere/bee/pkg/node.NewBee(0x154d468, 0x5, 0xc0001d0200, 0x20, 0x20, 0x187c6c0, 0x2194e80, 0xc0003a61e0, 0xc0003a62a0, 0x18786e0, ...)
	github.com/ethersphere/bee/pkg/node/node.go:316 +0xc9d
github.com/ethersphere/bee/cmd/bee/cmd.(*command).initStartCmd.func1(0xc0001662c0, 0xc0002dc300, 0x0, 0x6, 0x0, 0x0)
	github.com/ethersphere/bee/cmd/bee/cmd/start.go:118 +0x1138
github.com/spf13/cobra.(*Command).execute(0xc0001662c0, 0xc0002dc240, 0x6, 0x6, 0xc0001662c0, 0xc0002dc240)
	github.com/spf13/cobra@v1.0.0/command.go:842 +0x47c
github.com/spf13/cobra.(*Command).ExecuteC(0xc000166000, 0x0, 0x0, 0xc00004e2c0)
	github.com/spf13/cobra@v1.0.0/command.go:950 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.0.0/command.go:887
github.com/ethersphere/bee/cmd/bee/cmd.(*command).Execute(...)
	github.com/ethersphere/bee/cmd/bee/cmd/cmd.go:132
github.com/ethersphere/bee/cmd/bee/cmd.Execute(0xc0008b7f48, 0x44778a)
	github.com/ethersphere/bee/cmd/bee/cmd/cmd.go:141 +0x66
main.main()
	github.com/ethersphere/bee/cmd/bee/main.go:15 +0x26
```

File: pkg/localstore/migration.go [hunk 1]
```diff
@@ -101,3 +105,264 @@ func getMigrations(currentSchema, targetSchema string, allSchemeMigrations []mig
...
+		{"gcExclude", gcExcludeIndex},
+		{"retrievalAccess", retrievalAccessIndex},
+		{"retrievalData", retrievalDataIndex},
+	} {
+		db.logger.Infof("truncating %s index", v.name)
+		err := truncate(v.idx)
+		if err != nil {
+			return fmt.Errorf("truncate %s index: %w", v.name, err)
+		}
+		db.logger.Infof("truncated %d %s index entries", v.name, count)
```
Comment (Eknir) [inline pkg/localstore/migration.go:347]: I see on my terminal
```
INFO[2021-05-17T16:40:05+02:00] truncated %!d(string=pullsync) %!s(int=2) index entries 
```

Comment (Eknir) [reply pkg/localstore/migration.go:347]: Besides, I suggest to put all log lines to `debug` and only have two info log line saying that `migration is in process` and `migration successful`

File: pkg/localstore/migration.go [hunk 2]
```diff
@@ -101,3 +105,265 @@ func getMigrations(currentSchema, targetSchema string, allSchemeMigrations []mig
...
+	})
+	if err != nil {
+		return err
+	}
+	c, err := pinIndex.Count()
+	if err != nil {
+		return fmt.Errorf("get pin index count: %w", err)
+	}
+	if c > 0 {
+		return errors.New("pinned content detected. bee v0.6.0 cannot migrate a store with pinned content. please refer to the release notes for more information.")
```
Comment (Eknir) [inline pkg/localstore/migration.go:139]: This is very important information to the user, but contains a rather obscure and frightening message. I am making a suggestion below, mostly to give the user better instructions.

I would suggest to:
1. log an error message, saying: `failed to update your node due to the existence of pinned content. Please refer to the release notes on how to safely migrate your pinned content. Your localstore location is <path_to_localstore>`
2. return an error: `node cannot boot up with an outdated database scheme`

**Review by Eknir — APPROVED** [review]
LGTM. Tried migrating from v0.5.3 to v0.6.0 as a pinner and as a non-pinner

**Review by zelig — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1716)* status: all files reviewed, 4 unresolved discussions (waiting on @acud)

___
*[pkg/localstore/migration.go, line 135 at r3](https://reviewable.io/reviews/ethersphere/bee/1716#-M_x13LL0v2sStow3mQD:-M_x13LL0v2sStow3mQE:b2qva4l) ([raw file](https://github.com/ethersphere/bee/blob/e6b676db3ccd7c8a1fde5033ea6fca3709c9ab23/pkg/localstore/migration.go#L135)):*
> ```Go
> 	}
> 
> 	c, err := pinIndex.Count()
> ```

this is very inefficient to scan the entire index in order to count just to discover whether there is non-zero number of entries in the index. 
Just use the shed index iterator with a function that returns an error ("pinned content found") on any item, if the iterator call returns an error then return `fmt.Errorf("migration failed: %w.  Please refer to the release notes on how to safely migrate your pinned content. Your localstore location is %s", err, path_to_localstore)`   
___
*[pkg/localstore/migration.go, line 349 at r3](https://reviewable.io/reviews/ethersphere/bee/1716#-M_x2mCVBKRoX2mJssUs:-M_x2mCVBKRoX2mJssUt:b6o6r0t) ([raw file](https://github.com/ethersphere/bee/blob/e6b676db3ccd7c8a1fde5033ea6fca3709c9ab23/pkg/localstore/migration.go#L349)):*
> ```Go
> 			return fmt.Errorf("truncate %s index: %w", v.name, err)
> 		}
> 		db.logger.Debugf("truncated %d %s index entries", count, v.name)
> ```

`	db.logger.Debugf("truncated %d %s index entries", v.name,count)`

**Review by acud — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1716)* status: all files reviewed, 4 unresolved discussions (waiting on @Eknir and @zelig)

___
*[pkg/localstore/migration.go, line 347 at r1](https://reviewable.io/reviews/ethersphere/bee/1716#-M_fK0eO6wRC0AXH5nGu-r1-347:-M_y7DCJ-qTatSZQZBYt:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/3a356f8f717266fa3efd730e2cd966b8b3c0fda1/pkg/localstore/migration.go#L347)):*
<details><summary><i>Previously, Eknir (Eknir) wrote…</i></summary><blockquote>

Besides, I suggest to put all log lines to `debug` and only have two info log line saying that `migration is in process` and `migration successful`
</blockquote></details>

Done.
___
*[pkg/localstore/migration.go, line 139 at r2](https://reviewable.io/reviews/ethersphere/bee/1716#-M_fK0eO6wRC0AXH5nGu-r2-139:-M_y76gL-AM_DzqlgT4c:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/a6e29cae037a3c04713a1fdc55c584c5a9a7ab0d/pkg/localstore/migration.go#L139)):*
<details><summary><i>Previously, Eknir (Eknir) wrote…</i></summary><blockquote>

This is very important information to the user, but contains a rather obscure and frightening message. I am making a suggestion below, mostly to give the user better instructions.

I would suggest to:
1. log an error message, saying: `failed to update your node due to the existence of pinned content. Please refer to the release notes on how to safely migrate your pinned content. Your localstore location is <path_to_localstore>`
2. return an error: `node cannot boot up with an outdated database scheme`
</blockquote></details>

Done.
___
*[pkg/localstore/migration.go, line 135 at r3](https://reviewable.io/reviews/ethersphere/bee/1716#-M_x13LL0v2sStow3mQD:-M_y737a0Oi0LbbD0qAA:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/e6b676db3ccd7c8a1fde5033ea6fca3709c9ab23/pkg/localstore/migration.go#L135)):*
<details><summary><i>Previously, zelig (Viktor Trón) wrote…</i></summary><blockquote>

this is very inefficient to scan the entire index in order to count just to discover whether there is non-zero number of entries in the index. 
Just use the shed index iterator with a function that returns an error ("pinned content found") on any item, if the iterator call returns an error then return `fmt.Errorf("migration failed: %w.  Please refer to the release notes on how to safely migrate your pinned content. Your localstore location is %s", err, path_to_localstore)`   
</blockquote></details>

Done.
___
*[pkg/localstore/migration.go, line 349 at r3](https://reviewable.io/reviews/ethersphere/bee/1716#-M_x2mCVBKRoX2mJssUs:-M_y7EIE1Il_8pdHITSc:b-mgeeka) ([raw file](https://github.com/ethersphere/bee/blob/e6b676db3ccd7c8a1fde5033ea6fca3709c9ab23/pkg/localstore/migration.go#L349)):*
<details><summary><i>Previously, zelig (Viktor Trón) wrote…</i></summary><blockquote>

`	db.logger.Debugf("truncated %d %s index entries", v.name,count)`
</blockquote></details>

leaving like it is... `truncated x pullsync index entries`. not sure what's the problem with the text :shrug:

**Review by janos — APPROVED** [review]
> Reviewed 1 of 2 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1716)* status: all files reviewed, 4 unresolved discussions (waiting on @Eknir and @zelig)

**Review by zelig — APPROVED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1716)* status: all files reviewed, 2 unresolved discussions (waiting on @Eknir)

**Pull request merged**